tests/integrationtest/t/executor/cte.test

```diff
@@ -153,3 +153,19 @@ create table t1(c1 int);
 insert into t1 values(0), (1), (2), (3), (4), (5), (6), (7), (8), (9), (10), (11), (12), (13), (14), (15), (16), (17), (18), (19), (20), (21), (22), (23), (24), (25), (26), (27), (28), (29), (30), (31), (32), (33), (34), (35), (36), (37), (38), (39), (40), (41), (42), (43), (44), (45), (46), (47), (48), (49), (50), (51), (52), (53), (54), (55), (56), (57), (58), (59), (60), (61), (62), (63), (64), (65), (66), (67), (68), (69), (70), (71), (72), (73), (74), (75), (76), (77), (78), (79), (80), (81), (82), (83), (84), (85), (86), (87), (88), (89), (90), (91), (92), (93), (94), (95), (96), (97), (98), (99), (100), (101), (102), (103), (104), (105), (106), (107), (108), (109), (110), (111), (112), (113), (114), (115), (116), (117), (118), (119), (120), (121), (122), (123), (124), (125), (126), (127), (128), (129), (130), (131), (132), (133), (134), (135), (136), (137), (138), (139), (140), (141), (142), (143), (144), (145), (146), (147), (148), (149), (150), (151), (152), (153), (154), (155), (156), (157), (158), (159), (160), (161), (162), (163), (164), (165), (166), (167), (168), (169), (170), (171), (172), (173), (174), (175), (176), (177), (178), (179), (180), (181), (182), (183), (184), (185), (186), (187), (188), (189), (190), (191), (192), (193), (194), (195), (196), (197), (198), (199), (200), (201), (202), (203), (204), (205), (206), (207), (208), (209), (210), (211), (212), (213), (214), (215), (216), (217), (218), (219), (220), (221), (222), (223), (224), (225), (226), (227), (228), (229), (230), (231), (232), (233), (234), (235), (236), (237), (238), (239), (240), (241), (242), (243), (244), (245), (246), (247), (248), (249), (250), (251), (252), (253), (254), (255), (256), (257), (258), (259), (260), (261), (262), (263), (264), (265), (266), (267), (268), (269), (270), (271), (272), (273), (274), (275), (276), (277), (278), (279), (280), (281), (282), (283), (284), (285), (286), (287), (288), (289), (290), (291), (292), (293), (294), (295), (296), (297), (298), (299);
 with recursive cte1(c1) as (select c1 from t1 union select c1 + 1 c1 from cte1 limit 1 offset 100) select * from cte1;
 set tidb_max_chunk_size=default;
+
+# TestIssue46522
```

Comment:
I cannot reproduce the original panic in the issue with this ticket.

Reply:
![image](https://github.com/pingcap/tidb/assets/1420062/3c95ed5d-d67f-412c-bce8-324c7b6add13)

This is the error stack, 

```
[err="type assertion for CTEStorageMap failed
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).buildCTE
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/builder.go:5277
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).build
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/builder.go:308
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).buildInsert
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/builder.go:976
github.com/pingcap/tidb/pkg/executor.(*executorBuilder).build
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/builder.go:194
github.com/pingcap/tidb/pkg/executor.(*ExecStmt).buildExecutor
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/adapter.go:1203
github.com/pingcap/tidb/pkg/executor.(*ExecStmt).Exec
	/home/genius/project/src/github.com/pingcap/tidb/pkg/executor/adapter.go:550
github.com/pingcap/tidb/pkg/session.(*session).retry
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:1177
github.com/pingcap/tidb/pkg/session.(*session).doCommitWithRetry
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:898
github.com/pingcap/tidb/pkg/session.(*session).CommitTxn
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:1001
github.com/pingcap/tidb/pkg/session.autoCommitAfterStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/tidb.go:298
github.com/pingcap/tidb/pkg/session.finishStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/tidb.go:260
github.com/pingcap/tidb/pkg/session.runStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:2413
github.com/pingcap/tidb/pkg/session.(*session).ExecuteStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/session/session.go:2236
github.com/pingcap/tidb/pkg/server.(*TiDBContext).ExecuteStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/driver_tidb.go:293
github.com/pingcap/tidb/pkg/server.(*clientConn).handleStmt
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/conn.go:2045
github.com/pingcap/tidb/pkg/server.(*clientConn).handleQuery
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/conn.go:1835
github.com/pingcap/tidb/pkg/server.(*clientConn).dispatch
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/conn.go:1322
github.com/pingcap/tidb/pkg/server.(*clientConn).Run
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/conn.go:1101
github.com/pingcap/tidb/pkg/server.(*Server).onConn
	/home/genius/project/src/github.com/pingcap/tidb/pkg/server/server.go:701
runtime.goexit
	/home/genius/project/go/src/runtime/asm_amd64.s:1650
previous statement: insert into issue46522 with t1 as (select id+1 from  issue46522 where id = 1) select * from t1"]
```

I think they are the same type. @jackysp 
doCommitWithRetry() -> retry() -> ... -> buildCTE() -> type assertion for CTEStorageMap failed